importer.go

```diff
@@ -122,28 +122,56 @@ func (i *Importer) readViaCache(imp string) (jsonnet.Contents, error) {
 		return content, nil
 	}
 
-	u, err := addScheme(imp)
+	content, err := i.fetch(imp)
 	if err != nil {
 		return noContent, err
 	}
-	content, err := fetch(u, i.fetcher())
-	if err == nil {
-		i.cache[imp] = content
+
+	i.cache[imp] = content
+	return content, nil
+}
+
+func (i *Importer) fetch(imp string) (jsonnet.Contents, error) {
+	r, err := i.open(imp)
+	if r == nil || err != nil {
+		return noContent, err
 	}
 
-	return content, err
+	defer r.Close()
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return noContent, err
+	}
+
+	return jsonnet.MakeContents(string(b)), nil
 }
 
-func (i *Importer) fetcher() URLFetcher {
-	if i.Fetcher == nil {
-		// Create a http.Client with a transport handling file:// urls
-		// https://golang.org/pkg/net/http/#NewFileTransport
-		t := &http.Transport{}
-		t.RegisterProtocol("file", http.NewFileTransport(http.Dir("/")))
+func (i *Importer) open(imp string) (io.ReadCloser, error) {
+	if !isNetpath(imp) {
+		r, err := os.Open(imp)
+		if os.IsNotExist(err) {
+			return nil, nil
```

Comment:
why no error on not exist? I assume this is to do with:
```
// TODO(camh): Keep searching on hard errors. If a fetch results
		// in a hard error, then no amount of fetching that resource
		// will ever work with the path we have.
```
tbh, that doesn't make a heap of sense to me.
I assume IsNotExist is considered a hard error, yes?
How about it exists but you don't have permissions to read, is that less hard?
what if later I place the file that I want to import from in that location I originally wanted to import it from (i just typoed it before)? How is that different from a "transient" network failure?

The point is: imo errors are errors. To try and classify them into soft and hard feels like opening a can of worms.

Reply:
> The point is: imo errors are errors

I guess this is where we disagree.

First, IsNotExist is not an error that terminates the searching. I would not consider it an error at all from the point of view of the importer. The point of a search path is that you expect not to find imports in all places, so given that expectation, it's not an error unless the import does not exist anywhere on the search path. However, when reading an individual import, it will often get a IsNotExist. This is converted into the `nil` `io.Reader` or `noContent` (depending on the return type).

All other errors terminate the search and are not cached. But I think that if an attempt to retrieve an import can never work (such as a 400 Bad Request), then that should be cached and move on to the next search. It's not much different to IsNotExist.

But then there are transient errors. You don't want to get different results on different runs if you can help it. So a transient error should probably cause the import to fail so the whole evaluation fails. I get that the world can change between different runs, but that is usually the result of something deliberate and perhaps traceable.

The standard jsonnet importer fails on any non-IsNotExist error, so I have modelled mine on that, but I think with some http errors being permanent errors (usually 400 errors), there is a case to treat them like the IsNotExist error. Hence the TODO.

Reply:
well thanks for explaining your thoughts.